Modify rule S2259: LaYC format (#2294)
SonarSource/rspec: rules/S2259/dotnet-why.adoc

== Why is this an issue?

Accessing a {null-keyword-link}[{null-keyword}] value will always throw a https://learn.microsoft.com/en-us/dotnet/api/system.nullreferenceexception[NullReferenceException] most likely causing the application to terminate abruptly.

Such termination might expose sensitive information that a malicious third party could exploit to, for instance, bypass security measures.

=== Exceptions

> There is some scenario to which this rule won't raise:

Reviewer: ```suggestion
In the following cases, the rule does not raise:
```

* calls to extension methods are not reported because they can still operate on `{null-keyword}` values.
+
include::{exceptionCode1}[]

> * value has been previously validated by creating a `ValidatedNotNull` or `NotNull` attribute and used on method parameter

Reviewer: ```suggestion
* value has been validated by a helper method annotated with a `ValidatedNotNull` attribute
```

Reviewer: There are more options on how to suppress the warning:
These attributes are available in .Net core only but can be redefined in custom code and will be recognized by us:
* [NotNullAttribute](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.notnullattribute?view=net-7.0) same behavior as ValidatedNotNull
* NotNullWhen [see](https://github.com/SonarSource/sonar-dotnet/blob/8fc3b67ae9e293bf0ed77f68aa4de9fdfe3294ab/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.NetCore.cs#L92) same behavior as described in the [docs](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.notnullwhenattribute?view=net-7.0) 
* DoesNotReturn [see](https://github.com/SonarSource/sonar-dotnet/blob/8fc3b67ae9e293bf0ed77f68aa4de9fdfe3294ab/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp8.cs#L209) and [TerminatesProgram](https://github.com/SonarSource/sonar-dotnet/blob/8fc3b67ae9e293bf0ed77f68aa4de9fdfe3294ab/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp8.cs#L212C40-L212C40) Same behavior as described in the [docs](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.doesnotreturnattribute?view=net-7.0) and [here](https://blog.jetbrains.com/dotnet/2012/08/15/contract-annotations-in-resharper-7/)
* [DoesNotReturnIfAttribute](https://github.com/SonarSource/sonar-dotnet/blob/8fc3b67ae9e293bf0ed77f68aa4de9fdfe3294ab/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs#L1687) behaves like described in the [docs](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.codeanalysis.doesnotreturnifattribute?view=net-7.0)

* Debug.Assert: [see](https://github.com/SonarSource/sonar-dotnet/blob/8fc3b67ae9e293bf0ed77f68aa4de9fdfe3294ab/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.cs#L375)

We also support the bang operator:

* Bang operator [see](https://github.com/SonarSource/sonar-dotnet/blob/8fc3b67ae9e293bf0ed77f68aa4de9fdfe3294ab/analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/NullPointerDereference.CSharp8.Nullable.cs#L20)

Reviewer: The test case comment is wrong. That is a TP as `o1` is known to be null at that point.
